### DISCUSSION TITLE
q-colorencoding-server

### DISCUSSION THREAD

**Discussion by sanand0** [sanand0/tools-in-data-science-public#352] [Q&A]
Discussion for question `q-colorencoding-server` which appeared first in https://exam.sanand.workers.dev/tds-2026-01-ga7#hq-colorencoding-server

**Comment by Pratibimb-Swain** [comment]
Sir, the deadline for GA7 was April 5, but for me, it is showing the deadline was 29th March

**Reply by KrishayIITMBS** [reply]
same

**Reply by a-r-y-a-n09** [reply]
it has to be a bug don't worry guys it will be resolved facing the same issue

**Reply by AagmanChandra** [reply]
Same problem with me as well and we cannot check answer until this is resolved

**Reply by 23f3002599-sh** [reply]
can anyone access GA 8 it is showing it ended on April 8

**Comment by Pratibimb-Swain** [comment]
Is there anyone who got this correct?

**Reply by MASD-1045** [reply]
Yes please tell if anyone got this question right

**Reply by KrishayIITMBS** [reply]
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Elevation Change from Sea Level Reference</title>
  <script src="https://cdn.jsdelivr.net/npm/chart.js@4/dist/chart.umd.min.js"></script>
  <style>
    :root {
      color-scheme: light dark;
      --bg: white;
      --text: rgb(33, 37, 41);
      --subtext: rgb(108, 117, 125);
    }
    @media (prefers-color-scheme: dark) {
      :root {
        --bg: rgb(33, 37, 41);
        --text: rgb(248, 249, 250);
        --subtext: rgb(173, 181, 189);
      }
    }
    body { font-family: sans-serif; margin: 16px; background: var(--bg); color: var(--text); }
    h2 { font-size: 1rem; margin-bottom: 4px; }
    p  { font-size: 0.8rem; color: var(--subtext); margin-bottom: 12px; }
    canvas { max-height: 280px; }
  </style>
</head>
<body>
  <h2>Elevation Change from Sea Level Reference</h2>
  <p>Terrain elevation change (m) relative to local sea-level reference, from -85 m to +210 m</p>

  <canvas id="chart"></canvas>
  
  <script>
    /* diverging color scheme */
    /* EXPLANATION: one-directional ramp makes below-sea-level zones appear as low-elevation positive values, hiding that they are below the reference */

    const colors = ["#d73027", "#fdae61", "#ffffff", "#d9ef8b", "#a6d96a", "#66bd63", "#1a9850"];

    new Chart(document.getElementById('chart'), {
      type: 'bar',
      data: {
        labels: ["Zone A","Zone B","Zone C","Zone D","Zone E","Zone F","Zone G"],
        datasets: [{
          label: 'Elevation Change from Sea Level Reference',
          data: [-85,-30,0,25,70,140,210],
          backgroundColor: colors,
          borderColor: colors,
          borderWidth: 1
        }]
      },
      options: {
        responsive: true,
        plugins: { legend: { display: false } },
        scales: { y: { beginAtZero: false } }
      }
    });
  </script>
</body>
</html>

**Reply by KrishayIITMBS** [reply]
> <title>Elevation Change from Sea Level Reference</title> <script src="https://cdn.jsdelivr.net/npm/chart.js@4/dist/chart.umd.min.js"></script> <style> :root { color-scheme: light dark; --bg: white; --text: rgb(33, 37, 41); --subtext: rgb(108, 117, 125); } @media (prefers-color-scheme: dark) { :root { --bg: rgb(33, 37, 41); --text: rgb(248, 249, 250); --subtext: rgb(173, 181, 189); } } body { font-family: sans-serif; margin: 16px; background: var(--bg); color: var(--text); } h2 { font-size: 1rem; margin-bottom: 4px; } p { font-size: 0.8rem; color: var(--subtext); margin-bottom: 12px; } canvas { max-height: 280px; } </style>
> ## Elevation Change from Sea Level Reference
> Terrain elevation change (m) relative to local sea-level reference, from -85 m to +210 m
> 
> <script> /* diverging color scheme */ /* EXPLANATION: one-directional ramp makes below-sea-level zones appear as low-elevation positive values, hiding that they are below the reference */ const colors = ["#d73027", "#fdae61", "#ffffff", "#d9ef8b", "#a6d96a", "#66bd63", "#1a9850"]; new Chart(document.getElementById('chart'), { type: 'bar', data: { labels: ["Zone A","Zone B","Zone C","Zone D","Zone E","Zone F","Zone G"], datasets: [{ label: 'Elevation Change from Sea Level Reference', data: [-85,-30,0,25,70,140,210], backgroundColor: colors, borderColor: colors, borderWidth: 1 }] }, options: { responsive: true, plugins: { legend: { display: false } }, scales: { y: { beginAtZero: false } } } }); </script>

view the source code if you want the accurate code

**Reply by 23f2001258** [reply]
# This is the correct solution.
```HTML
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Website Traffic by Source</title>
  <script src="https://cdn.jsdelivr.net/npm/chart.js@4/dist/chart.umd.min.js"></script>
</head>
<body style="font-family: sans-serif; margin: 16px;">

  <!--
    SCHEME TYPE: categorical

    EXPLANATION:
    The original chart applied a sequential single-hue ramp to unordered nominal groups.
    This sequential ramp falsely implies traffic sources have a natural progression or hierarchy,
    misleading viewers into thinking that darker bars indicate a higher tier, rank, or progression.
    Because website traffic sources (Organic, Paid, Social, Direct, Email) are independent categories
    with no intrinsic sequence, a categorical color scheme must be used.

    We apply a perceptually distinct categorical palette:
    ["#e41a1c", "#377eb8", "#4daf4a", "#984ea3", "#ffbb00"]
    All pairwise CIEDE2000 distances are strictly >= 30, ensuring full discriminability.
  -->

  <h2>Website Traffic by Source</h2>
  <p>Monthly visits (thousands) by traffic source: Organic, Paid, Social, Direct, Email</p>
  <canvas id="chart" style="max-height: 280px;"></canvas>

  <script>
    // Corrected categorical palette with all pairwise CIEDE2000 distances >= 30
    const colors = ["#e41a1c", "#377eb8", "#4daf4a", "#984ea3", "#ffbb00"];

    new Chart(document.getElementById('chart'), {
      type: 'bar',
      data: {
        labels: ["Organic", "Paid", "Social", "Direct", "Email"],
        datasets: [{
          label: 'Website Traffic by Source',
          data: [85, 42, 33, 67, 18],
          backgroundColor: colors,
          borderColor: colors,
          borderWidth: 1
        }]
      },
      options: {
        responsive: true,
        plugins: { legend: { display: false } },
        scales: { y: { beginAtZero: true } }
      }
    });
  </script>
</body>
</html>
```

**Reply by 23f2001258** [reply]
For this question, I used the following notebook.
[colormap.ipynb](https://github.com/user-attachments/files/32695367/colormap.ipynb)
- In Terminal
```bash
pip install numpy scikit-learn scikit-image matplotlib seaborn
```

- Or, in Python Notebook
```python
!pip install numpy scikit-learn scikit-image matplotlib seaborn
```
- Then, ---
```python
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from skimage.color import rgb2lab, deltaE_ciede2000
from sklearn.manifold import MDS

hex_colors = ["#e41a1c", "#377eb8", "#4daf4a", "#984ea3", "#ffbb00"]

# 1. Convert Hex to CIELAB
def hex_to_lab(hex_str):
    rgb = np.array([int(hex_str.lstrip('#')[i:i+2], 16) for i in (0, 2, 4)]) / 255.0
    return rgb2lab(rgb.reshape(1, 1, 3)).reshape(3)

lab_colors = np.array([hex_to_lab(c) for c in hex_colors])
n = len(hex_colors)

# 2. Build CIEDE2000 Distance Matrix
dist_matrix = np.zeros((n, n))
for i in range(n):
    for j in range(n):
        if i != j:
            dist_matrix[i, j] = deltaE_ciede2000(lab_colors[i], lab_colors[j])

# 3. Project distances onto a 2D Map using Multidimensional Scaling (MDS)
mds = MDS(n_components=2, dissimilarity='precomputed', random_state=42)
coords_2d = mds.fit_transform(dist_matrix)

# 4. Plot both views
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(14, 5.5))

# (Left) Distance Heatmap
sns.heatmap(dist_matrix, annot=True, fmt=".1f", xticklabels=hex_colors,
            yticklabels=hex_colors, cmap="Blues", ax=ax1, cbar_kws={'label': 'CIEDE2000 Distance'})
ax1.set_title("Pairwise Distance Matrix (Need ≥ 30)")

# (Right) 2D Perceptual Map
for i, hex_code in enumerate(hex_colors):
    x, y = coords_2d[i]
    ax2.scatter(x, y, color=hex_code, s=400, edgecolors='black', linewidths=1.5, zorder=3)
    ax2.text(x, y + 2.5, hex_code, ha='center', fontsize=9, fontweight='bold')

ax2.grid(True, linestyle='--', alpha=0.5)
ax2.set_title("2D MDS Map (Physical Distance = Perceptual Difference)")
ax2.set_xlabel("MDS Dimension 1")
ax2.set_ylabel("MDS Dimension 2")

plt.tight_layout()
plt.show()
```
